### PR TITLE
Add a cooldown to /report, resolves Electroid#242

### LINKED ISSUE #242: Report cooldown
OCN had this. It's better to prevent report spamming with a short cooldown instead of having to kick or ban players who spam the command.

### PR DESCRIPTION
**Pull request by applenick**
Resolves #242 

The changes were inspired by [here](https://github.com/OvercastNetwork/ProjectAres/blob/b883ef579938c49076379fa306592da847fc672f/Commons/bukkit/src/main/java/tc/oc/commons/bukkit/report/ReportCommands.java#L105) 

I did not want to deviate too far from what has been done in the past. The cooldown is implemented with a hard-coded 15 second cooldown, which I feel will be appropriate for most situations.

 If it is desired I can add a config option or adjust the hard-coded limit, just let me know.

~~Edit: Forgot to mention, I did not add a bypass for players with a specific permission. As I can’t really imagine any situation where one would want to bypass this check. Anyway, if that is desired I can happily add it too.~~ (Outdated)

Screenshot of formatting (what would happen if user spams the command several times):
Screen Shot 2020-01-17 at 10 40 32 AM
Signed-off-by: applenick <applenick@users.noreply.github.com>

### CODE REVIEW

**Comment by Meeples10** [comment]
There should probably be a config option for this that defaults to 15 seconds.

File: src/main/java/tc/oc/pgm/commands/ModerationCommands.java [hunk 1]
```diff
@@ -18,16 +24,40 @@
 
 public class ModerationCommands {
 
+  private static final int REPORT_COOLDOWN_SECONDS = 15;
+
+  private final Map<CommandSender, Instant> lastReportSent = new WeakHashMap<>();
```
Comment (Pablete1234) [inline src/main/java/tc/oc/pgm/commands/ModerationCommands.java]: Should use a guava expiring cache, and maybe just use the UUID as key (if the command sender is the console just always allow)

**Comment by Pablete1234** [comment]
I think the 15 second constant is fine, there's no need to over-complicate such a simple feature

File: src/main/java/tc/oc/pgm/commands/ModerationCommands.java (hunk 1 above)
Comment (Electroid) [reply src/main/java/tc/oc/pgm/commands/ModerationCommands.java]: Also exempt players with the `Permissions.STAFF`

Comment (applenick) [reply src/main/java/tc/oc/pgm/commands/ModerationCommands.java]: Thanks @Pablete1234 and @Electroid! 
I switched the  map over to use a guava loading cache. I haven’t used them before, so hopefully the way I implemented it is the proper way. Though after testing everything seemed to work, so it should be right 😄 

Also I added the staff bypass.

**Comment by applenick** [comment]
Thanks @Pablete1234 and @Electroid!
I switched the map over to use a guava loading cache. I haven’t used them before, so hopefully the way I implemented it is the proper way. Though after testing everything seemed to work, so it should be right 😄

I also agree with @Pablete1234 that this is a very simple feature and really there is no reason to add an additional config option.

File: src/main/java/tc/oc/pgm/commands/ModerationCommands.java [hunk 2]
```diff
@@ -18,6 +26,14 @@
 
 public class ModerationCommands {
 
+  private static final int REPORT_COOLDOWN_SECONDS = 15;
+
+  private static final Cache<UUID, Instant> lastReportSent =
+      CacheBuilder.newBuilder()
+          .weakKeys()
```
Comment (Pablete1234) [inline src/main/java/tc/oc/pgm/commands/ModerationCommands.java]: Using uuis, no need for weak keys

**Comment by Pugzy** [comment]
Would it be better to make the translated message more generic?
> Please wait {0} before running that command again

Allowing this to be extracted to more commands in the future

File: src/main/java/tc/oc/pgm/commands/ModerationCommands.java (hunk 2 above)
Comment (applenick) [reply src/main/java/tc/oc/pgm/commands/ModerationCommands.java]: Ah thanks, I was unsure! I’ll adjust that now

**Comment by applenick** [comment]
> Would it be better to make the translated message more generic?
> 
> > Please wait {0} before running that command again
> 
> Allowing this to be extracted to more commands in the future

Great suggestion. I’ll adjust that as well.

File: src/main/java/tc/oc/pgm/commands/ModerationCommands.java [hunk 3]
```diff
@@ -27,7 +40,32 @@ public static void report(
...
+        if (secondsRemaining > 0) {
+          throw new CommandException(
+              AllTranslations.get()
+                  .translate(
+                      "command.cooldown",
+                      commandSender,
+                      ChatColor.AQUA
+                          + (secondsRemaining
+                              + " second"
+                              + (secondsRemaining != 1 ? "s" : "")
```
Comment (TheMolkaPL) [inline src/main/java/tc/oc/pgm/commands/ModerationCommands.java]: This should be translatable.

Comment (applenick) [reply src/main/java/tc/oc/pgm/commands/ModerationCommands.java]: Thanks Molka, I should have done this from the beginning!

**Comment by TheMolkaPL** [comment]
I meant something like this: :P

https://github.com/Electroid/PGM/blob/31ebbcc74002981fcf05dc0af45e861a77d07aca/src/main/i18n/templates/strings.properties#L672-L674

File: src/main/java/tc/oc/pgm/commands/ModerationCommands.java [hunk 4]
```diff
@@ -18,6 +25,11 @@
 
 public class ModerationCommands {
 
+  private static final int REPORT_COOLDOWN_SECONDS = 15;
+
+  private static final Cache<UUID, Instant> lastReportSent =
+      CacheBuilder.newBuilder().expireAfterWrite(REPORT_COOLDOWN_SECONDS, TimeUnit.SECONDS).build();
```
Comment (TheMolkaPL) [inline src/main/java/tc/oc/pgm/commands/ModerationCommands.java]: I think this should be called `LAST_REPORT_SENT`.

Comment (applenick) [reply src/main/java/tc/oc/pgm/commands/ModerationCommands.java]: > I meant something like this: :P
> 
> https://github.com/Electroid/PGM/blob/31ebbcc74002981fcf05dc0af45e861a77d07aca/src/main/i18n/templates/strings.properties#L672-L674

Whoops! 😆 
I’ll make the adjustments now, thanks again haha.

File: src/main/java/tc/oc/pgm/commands/ModerationCommands.java [hunk 5]
```diff
@@ -27,7 +39,36 @@ public static void report(
...
+      @Text String reason)
+      throws CommandException {
+    if (!commandSender.hasPermission(Permissions.STAFF) && commandSender instanceof Player) {
+      // Check for cooldown
+      Instant lastReport = LAST_REPORT_SENT.getIfPresent(matchPlayer.getId());
+      if (lastReport != null) {
+        Duration timeSinceReport = Duration.between(lastReport, Instant.now());
+        long secondsRemaining = REPORT_COOLDOWN_SECONDS - timeSinceReport.getSeconds();
+        if (secondsRemaining > 0) {
+          Component secondsComponent = new PersonalizedText("" + secondsRemaining);
```
Comment (TheMolkaPL) [inline src/main/java/tc/oc/pgm/commands/ModerationCommands.java]: I would use `Integer.toString(secondsRemaining)`, otherwise looks good to me.

Comment (applenick) [reply src/main/java/tc/oc/pgm/commands/ModerationCommands.java]: Oh snap, I did not see your suggestion! I’m adjusting this right now. 👍

**Pull request merged**